src/Razor/Microsoft.CodeAnalysis.Razor/test/DefaultTagHelperDescriptorProviderTest.cs

```diff
@@ -50,5 +50,99 @@ public void Execute_NoOpsIfCompilationIsNotSet()
             // Assert
             Assert.Empty(context.Results);
         }
+
+        [Fact]
+        public void Execute_WithFilterAssemblyDoesNotDiscoverTagHelpersFromReferences()
+        {
+            // Arrange
+            var typeName = "TestAssembly.TestTagHelper";
+            var csharp = @"
+using Microsoft.AspNetCore.Razor.TagHelpers;
+namespace TestAssembly
+{
+    public class TestTagHelper : TagHelper
+    {
+        public override void Process(TagHelperContext context, TagHelperOutput output) {}
+    }
+}";
+            var compilation = TestCompilation.Create(_assembly, CSharpSyntaxTree.ParseText(csharp));
+            var descriptorProvider = new DefaultTagHelperDescriptorProvider();
+
+            var context = TagHelperDescriptorProviderContext.Create();
+            context.SetCompilation(compilation);
+            context.Items.SetTagHelperDiscoveryFilter(TagHelperDiscoveryFilter.CurrentCompilation);
+
+            // Act
+            descriptorProvider.Execute(context);
+
+            // Assert
+            Assert.NotNull(compilation.GetTypeByMetadataName(typeName));
+            var descriptor = Assert.Single(context.Results);
+            Assert.Equal(typeName, descriptor.GetTypeName());
+        }
+
+        [Fact]
+        public void Execute_WithFilterReferenceDoesNotDiscoverTagHelpersFromAssembly()
+        {
+            // Arrange
+            var testTagHelper = "TestAssembly.TestTagHelper";
+            var enumTagHelper = "Microsoft.CodeAnalysis.Razor.Workspaces.Test.EnumTagHelper";
+            var csharp = @"
+using Microsoft.AspNetCore.Razor.TagHelpers;
+namespace TestAssembly
+{
+    public class TestTagHelper : TagHelper
+    {
+        public override void Process(TagHelperContext context, TagHelperOutput output) {}
+    }
+}";
+            var compilation = TestCompilation.Create(_assembly, CSharpSyntaxTree.ParseText(csharp));
+            var descriptorProvider = new DefaultTagHelperDescriptorProvider();
+
+            var context = TagHelperDescriptorProviderContext.Create();
+            context.SetCompilation(compilation);
+            context.Items.SetTagHelperDiscoveryFilter(TagHelperDiscoveryFilter.ReferenceAssemblies);
+
+            // Act
+            descriptorProvider.Execute(context);
+
+            // Assert
+            Assert.NotNull(compilation.GetTypeByMetadataName(testTagHelper));
+            Assert.NotEmpty(context.Results);
+            Assert.Empty(context.Results.Where(f => f.GetTypeName() == testTagHelper));
+            Assert.NotEmpty(context.Results.Where(f => f.GetTypeName() == enumTagHelper));
+        }
+
+        [Fact]
+        public void Execute_WithDefaultDiscoversTagHelpersFromAssemblyAndReference()
+        {
+            // Arrange
+            var testTagHelper = "TestAssembly.TestTagHelper";
+            var enumTagHelper = "Microsoft.CodeAnalysis.Razor.Workspaces.Test.EnumTagHelper";
```

Comment:
How does this `EnumTagHelper` come in to play? I would expect that there is a project reference to `Microsoft.CodeAnalysis.Razor.Workspaces.Test` somewhere but I am missing where that happens...

Reply:
The test compilation references the test project which [has these types](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Razor/Microsoft.CodeAnalysis.Razor/test/TagHelperDescriptorFactoryTagHelpers.cs). It kinda simulates having a project reference like Mvc or Components.